### PR TITLE
Prevent read procedures from writing

### PR DESCRIPTION
**Pull request by boggle**
This makes sure we inject a specialized GDS into procedures that controls which capabilities are available to them. To do so, it introduces a lightweight AccessLevel facility to KernelTransaction.
Additionally, the GDS instance injected into procedures cannot: open/close transactions, work with kernel hooks, and autoindexing at this time.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/AvailabilityGuard.java [hunk 1]
```diff
@@ -25,6 +25,7 @@
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Function;
 
+import org.neo4j.graphdb.DatabaseShutdownException;
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/AvailabilityGuard.java]: used?

File: community/kernel/src/main/java/org/neo4j/kernel/AvailabilityGuard.java [hunk 2]
```diff
@@ -72,6 +73,7 @@ public UnavailableException( String message )
         String description();
     }
 
+
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/AvailabilityGuard.java]: why?

File: community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java [hunk 3]
```diff
@@ -190,25 +201,25 @@ RelationshipIterator nodeGetRelationships( long nodeId,
      *
      * @throws org.neo4j.kernel.api.exceptions.index.IndexNotFoundKernelException if no such index found.
      */
-    long nodeGetFromUniqueIndexSeek( IndexDescriptor index, Object value ) throws IndexNotFoundKernelException,
+    long nodeGetFromUniqueIndexSeek(IndexDescriptor index, Object value) throws IndexNotFoundKernelException,
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java]: I believe the old formatting was the _correct_ one? (true for all changes in this file)

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListLabelsProcedure.java [hunk 4]
```diff
@@ -42,7 +42,7 @@ public ListLabelsProcedure( ProcedureName name )
     @Override
     public RawIterator<Object[], ProcedureException> apply( Context ctx, Object[] input ) throws ProcedureException
     {
-        RawIterator<Label,ProcedureException> labels = asRawIterator( TokenAccess.LABELS.inUse( ctx.get( statement ) ) );
+        RawIterator<Label,ProcedureException> labels = asRawIterator( TokenAccess.LABELS.inUse( ctx.get(STATEMENT) ) );
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListLabelsProcedure.java]: formatting: `ctx.get( STATEMENT )`

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListPropertyKeysProcedure.java [hunk 5]
```diff
@@ -41,7 +41,7 @@ public ListPropertyKeysProcedure( ProcedureSignature.ProcedureName name )
     @Override
     public RawIterator<Object[], ProcedureException> apply( Context ctx, Object[] input ) throws ProcedureException
     {
-        RawIterator<String,ProcedureException> labels = asRawIterator( TokenAccess.PROPERTY_KEYS.inUse( ctx.get( statement ) ) );
+        RawIterator<String,ProcedureException> labels = asRawIterator( TokenAccess.PROPERTY_KEYS.inUse( ctx.get(STATEMENT) ) );
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListPropertyKeysProcedure.java]: formatting

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListRelationshipTypesProcedure.java [hunk 6]
```diff
@@ -43,7 +43,7 @@ public ListRelationshipTypesProcedure( ProcedureSignature.ProcedureName name )
     public RawIterator<Object[], ProcedureException> apply( Context ctx, Object[] input ) throws ProcedureException
     {
         RawIterator<RelationshipType,ProcedureException> types =
-                asRawIterator( TokenAccess.RELATIONSHIP_TYPES.inUse( ctx.get( statement ) ) );
+                asRawIterator( TokenAccess.RELATIONSHIP_TYPES.inUse( ctx.get(STATEMENT) ) );
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/ListRelationshipTypesProcedure.java]: formatting

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java [hunk 7]
```diff
@@ -1395,21 +1393,28 @@ public DoubleLongRegister indexSample( IndexDescriptor index, DoubleLongRegister
...
+                ctx.put( ReadOperations.STATEMENT, statement );
+                return procedures.call( ctx, name, input );
+            }
+            finally
+            {
+                reset.resetAccessLevel();
+            }
         }
-        finally
+        catch ( UnauthorizedAccessKernelException e )
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java]: Maybe get rid one level of nesting here? Turn the inner try catch to a method.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java [hunk 8]
```diff
@@ -66,6 +66,7 @@
 import org.neo4j.kernel.api.properties.PropertyKeyIdIterator;
 import org.neo4j.kernel.api.txstate.TransactionCountingStateVisitor;
 import org.neo4j.kernel.api.txstate.TransactionState;
+import org.neo4j.kernel.api.index.AutoIndexing;
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java]: remove

File: community/neo4j-harness/neo4j-home/ssl/snakeoil.cert [hunk 9]
```diff
@@ -0,0 +1,11 @@
+-----BEGIN CERTIFICATE-----
```
Comment (pontusmelke) [inline community/neo4j-harness/neo4j-home/ssl/snakeoil.cert]: remove

File: community/neo4j-harness/neo4j-home/ssl/snakeoil.key [hunk 10]
```diff
@@ -0,0 +1,16 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (pontusmelke) [inline community/neo4j-harness/neo4j-home/ssl/snakeoil.key]: remove

File: community/server/neo4j-home/ssl/snakeoil.cert [hunk 11]
```diff
@@ -0,0 +1,11 @@
+-----BEGIN CERTIFICATE-----
```
Comment (pontusmelke) [inline community/server/neo4j-home/ssl/snakeoil.cert]: remove

File: community/server/neo4j-home/ssl/snakeoil.key [hunk 12]
```diff
@@ -0,0 +1,16 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (pontusmelke) [inline community/server/neo4j-home/ssl/snakeoil.key]: remove

File: community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java [hunk 13]
```diff
@@ -58,8 +58,7 @@
 /**
  * Defines all types of read operations that can be done from the {@link KernelAPI}.
  */
-public interface ReadOperations
-{
+public interface ReadOperations {
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java]: This file is not formatted correctly

File: community/server/data/dbms/auth [hunk 14]
```diff
@@ -0,0 +1 @@
+neo4j:SHA-256,02D1DC324C275D63801047F3B37E3D47C7BFCCD1DE607BE433C5F64A3BFC7D28,AEF1BB85ACFB248E7F1BCA5FA216F067:password_change_required
```
Comment (pontusmelke) [inline community/server/data/dbms/auth]: remove

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/server/edge/EnterpriseEdgeEditionModule.java [hunk 15]
```diff
@@ -117,18 +118,14 @@ public EnterpriseEdgeEditionModule( final PlatformModule platformModule,
...
         headerInformationFactory = TransactionHeaderInformationFactory.DEFAULT;
 
-        schemaWriteGuard = new SchemaWriteGuard()
-        {
-            @Override
-            public void assertSchemaWritesAllowed() throws InvalidTransactionTypeKernelException
-            {
-            }
-        };
+        schemaWriteGuard = () -> {};
```
Comment (MishaDemianenko) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/server/edge/EnterpriseEdgeEditionModule.java]: There 3 implementations like that in a code base, that should be extracted and class should be introduced instead

File: community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/UnauthorizedAccessKernelException.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.exceptions;
+
+public class UnauthorizedAccessKernelException extends KernelException
+{
+    public UnauthorizedAccessKernelException( String message )
+    {
+        super( Status.Transaction.UnauthorizedAccess, (Throwable) null, message );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/UnauthorizedAccessKernelException.java]: why do we need to puss  `(Throwable) null`? There is another constructor without `Throwable`

File: community/kernel/src/main/java/org/neo4j/kernel/api/AccessLevel.java [hunk 17]
```diff
@@ -52,8 +55,17 @@ public boolean allowsSchemaWrites()
...
+
+        @Override
+        public boolean mayLimitAccessTo( AccessLevel newMode ) { return true; }
     };
 
     public abstract boolean allowsWrites();
     public abstract boolean allowsSchemaWrites();
+    public abstract boolean mayLimitAccessTo( AccessLevel newMode );
+
+    public interface Reset
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/api/AccessLevel.java]: Does not sound like a good place for something that will reset level in KernelTransaction, looks strange and suspicious

**Comment by MishaDemianenko** [comment]
guys please avoid reformatting in a PRs that brings something else inside, it's really hard to follow

File: community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java [hunk 18]
```diff
@@ -561,15 +603,24 @@ DoubleLongRegister indexSample( IndexDescriptor index, DoubleLongRegister target
     //== PROCEDURE OPERATIONS ===================
     //===========================================
 
-    /** For managed procedures, this gives access to the current statement. */
-    CallableProcedure.Key<Statement> statement = CallableProcedure.Key.key("statement", Statement.class );
+    /**
+     * For managed procedures, this gives access to the current transaction.
+     */
+    CallableProcedure.Key<KernelTransaction> KERNEL_TRANSACTION = CallableProcedure.Key.key( "KernelTransaction",
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java]: i find it really strange that predefined Context keys defined elsewhere, can we move it there?

File: community/kernel/src/test/java/org/neo4j/kernel/GraphDatabaseFacadeFactoryTest.java [hunk 19]
```diff
@@ -143,7 +146,8 @@ public LifeSupport createLife()
             @Override
             protected EditionModule createEdition( PlatformModule platformModule )
             {
-                return null;
+                EditionModule edition = Mockito.mock( EditionModule.class, Answers.RETURNS_DEEP_STUBS.get() );
```
Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/GraphDatabaseFacadeFactoryTest.java]: please use `Mockito.RETURNS_DEEP_STUBS` instead

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+        this.availability = availability;
+    }
+
+    @Override
+    public GraphDatabaseService apply( CallableProcedure.Context context ) throws ProcedureException
+    {
+        KernelTransaction transaction = context.get( ReadOperations.KERNEL_TRANSACTION );
+
+        GraphDatabaseFacade facade = new GraphDatabaseFacade();
+        facade.init( config, new GraphDatabaseFacade.SPI()
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java]: this SPI implementation should be extracted into it's own class

File: community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexing.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.index;
+
+public interface AutoIndexing
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexing.java]: class level and method level java docs are missing

File: community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexing.java [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.index;
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexing.java]: since autoindexes is legacy indexes that should be reflected in package name

File: community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexOperations.java [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+package org.neo4j.kernel.api.index;
+
+import java.util.Collections;
+import java.util.Set;
+
+import org.neo4j.kernel.api.DataWriteOperations;
+import org.neo4j.kernel.api.exceptions.legacyindex.AutoIndexingKernelException;
+import org.neo4j.kernel.api.properties.Property;
+
+public interface AutoIndexOperations
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexOperations.java]: class level and method level javadocs

File: community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexOperations.java [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.index;
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/api/index/AutoIndexOperations.java]: package name should reflect legacy index stuff

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexing.java [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.index;
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexing.java]: legacy package

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.index;
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java]: package name should be changed to include legacy notion

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java [hunk 27]
```diff
@@ -0,0 +1,316 @@
...
+                throws LegacyIndexNotFoundKernelException, EntityNotFoundException;
+
+        public abstract void remove( DataWriteOperations ops, long entityId )
+                throws LegacyIndexNotFoundKernelException, EntityNotFoundException;
+
+        public abstract void ensureIndexExists( DataWriteOperations ops )
+                throws LegacyIndexNotFoundKernelException, EntityNotFoundException;
+    }
+
+    private AtomicReference<Set<String>> propertyKeysToInclude = new AtomicReference<>( Collections.emptySet() );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java]: if we will add propertyKeysToInclude to constructor we should be able to make it just a plain set

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java [hunk 28]
```diff
@@ -0,0 +1,316 @@
...
+            }
+            catch ( LegacyIndexNotFoundKernelException | EntityNotFoundException e )
+            {
+                throw new AutoIndexingKernelException( e );
+            }
+        }
+    }
+
+    // Trap door needed to keep this as an enum
+    void replacePropertyKeysToInclude( List<String> propertyKeysToIncludeNow )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java]: as soon as propertyKeysToInclude will be included in constructor this method should go.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java [hunk 29]
```diff
@@ -0,0 +1,316 @@
...
+                    type.add( ops, entityId, name, property.value() );
+                }
+            }
+            catch ( LegacyIndexNotFoundKernelException | EntityNotFoundException | PropertyNotFoundException e )
+            {
+                throw new AutoIndexingKernelException( e );
+            }
+            catch ( TokenNotFoundException e )
+            {
+                // TODO: TokenNotFoundException was added before there was a kernel. It should be converted to a
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java]: all TODOs in this class should be addressed

File: community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java [hunk 30]
```diff
@@ -164,9 +163,11 @@
         <T> void unregisterTransactionEventHandler( TransactionEventHandler<T> handler );
 
         URL validateURLAccess( URL url ) throws URLAccessValidationError;
+
+        String toFacadeString();
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java]: What is the purpose of this method and why it should be in `GraphDatabaseFacade`?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java [hunk 31]
```diff
@@ -0,0 +1,213 @@
...
+import org.neo4j.kernel.impl.query.QueryExecutionEngine;
+import org.neo4j.kernel.impl.query.QueryExecutionKernelException;
+import org.neo4j.kernel.impl.query.QuerySession;
+import org.neo4j.kernel.impl.store.StoreId;
+
+import java.net.URL;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class ProcedureGDSFactory implements ThrowingFunction<CallableProcedure.Context,GraphDatabaseService,ProcedureException>
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java]: Factory should implement `Factory` instead of `ThrowingFunction`

File: community/collections/src/main/java/org/neo4j/helpers/collection/Iterables.java [hunk 32]
```diff
@@ -33,6 +33,7 @@
 import java.util.function.Predicate;
 
 import org.neo4j.collection.RawIterator;
+import org.neo4j.function.ThrowingAction;
```
Comment (MishaDemianenko) [inline community/collections/src/main/java/org/neo4j/helpers/collection/Iterables.java]: redundant import

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java [hunk 33]
```diff
@@ -0,0 +1,316 @@
...
+    @Override
+    public boolean enabled()
+    {
+        return enabled;
+    }
+
+    @Override
+    public void startAutoIndexingProperty( String propName )
+    {
+        propertyKeysToInclude.getAndUpdate( current -> {
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java]: as soon as set will be released from atomicReference, this can be simply achieved by any concurrent set implementations if multi threaded access is required.

**Comment by MishaDemianenko** [comment]
@boggle should we also prevent people from starting new thread inside a procedures?

**Comment by MishaDemianenko** [comment]
@boggle finish my first review round; couple of comments and also needs rebase

File: community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java (hunk 18 above)
Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java]: +1

File: community/kernel/src/test/java/org/neo4j/kernel/GraphDatabaseFacadeFactoryTest.java (hunk 19 above)
Comment (boggle) [inline community/kernel/src/test/java/org/neo4j/kernel/GraphDatabaseFacadeFactoryTest.java]: Is there a reason for that? I thought both are basically equivalent and public API

Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/GraphDatabaseFacadeFactoryTest.java]: `Mockito` is recommended public facade class that should decrease number of mockito classes that user code interact with

**Comment by boggle** [comment]
On formatting: Sorry that happens whenever I'm still on the scala style and is mostly done by IDEA without me noticing. Would be awesome to have a code style file that works for both Java and scala.

**Comment by boggle** [comment]
On AutoIndexing: This is really just refactoring; I'm happy to move things into the right places but it's out of scope of this PR to address old TODOs. Happy to have a look at least.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java (hunk 33 above)
Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java]: Since these sets are updated very, very rarely, I believe for reading, using an immutable set obtained from an atomic reference is faster than using a concurrent set.

**Comment by boggle** [comment]
On preventing threads: No, we want them to be able to do that. Also achieving that is hard, you'd have to mess with classloaders to get that done or perhaps the Java security model.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java (hunk 31 above)
Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java]: Can't use that, 1) we have an argument per call to the factory and don't want to create it again and again (maybe we could do that) 2) we need to throw a checked exception, factory interface doesn't allow that

Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java]: Then it looks like you can use ThrowingSupplier for all of that

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java (hunk 33 above)
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/InternalAutoIndexOperations.java]: i was more concerned about code simplicity then speed here

**Comment by MishaDemianenko** [comment]
On Thread prevention: was thinking about custom security manager and thread group policies, since think that will bring more value.

**Comment by MishaDemianenko** [comment]
On Auto indexing: yep saw that, can we just move it to another PR then? or you should promise to do that in next PR)?

**Comment by boggle** [comment]
I had to rebuild all the commits here and tried to address your feedback.

**Comment by boggle** [comment]
@MishaDemianenko Ping
